### PR TITLE
Add document on persistence.

### PR DESCRIPTION
**Pull request by uglybugger**
This PR adds a write-up of our position on persistence within the Octopus Server codebase.

### CODE REVIEW

File: OctopusServer/Persistence.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+NOTE: Our intention is to revisit our long-running `TaskController` pattern to decrease the need for long-lived, in-memory state, but that's outside the scope of this document.
+
+Within a single unit of work, we might touch several different entity types from several different projects. Each of those projects will have
+its own persistence settings:
+
+- Some projects might be wholly in the database.
+- Some projects might be in their own Git repository.
+- Some projects might be in a Git monorepository shared with several other projects.
+
+All of this means that there is no one "transaction" to commit.
```
Comment (Waldo000000) [inline OctopusServer/Persistence.md:24]: Great point. Another FAQ we might expect: Is the UoW atomic? i.e. What happens if one transaction successfully commits but another (in the same UoW) fails?

File: OctopusServer/Persistence.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+We wrap a logical unit of work around this collection of (0-many) individual transactions. The unit of work is completed or abandoned automatically
+via aspects (filters, middleware etc.) and should not need to be addressed directly during the normal course of operations.
+
+# Custom queries
+
+The `IDocumentStore<TDocument>` interfaces exposes a `.Query()` method which returns an `IQueryable<TDocument>`. The expected approach to writing a custom query is to use a LINQ expression against that `IQueryable<T>`. We have implementations of `IQueryable<T>` for both SQL and Git.
```
Comment (Waldo000000) [inline OctopusServer/Persistence.md:31]:
```suggestion
The `IDocumentStore<TDocument>` interface exposes a `.Query()` method which returns an `IQueryable<TDocument>`. The expected approach to writing a custom query is to use a LINQ expression against that `IQueryable<T>`. We have implementations of `IQueryable<T>` for both SQL and Git.
```

File: OctopusServer/Persistence.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+Where a custom query is reused in multiple locations, we use an extension method class on the `IDocumentStore<TDocument>` interface. For example, we have a [reusable mechanism for loading a `Project` via its slug](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Core/Features/Projects/ProjectDocumentStoreExtensionMethods.cs).
```
Comment (Waldo000000) [inline OctopusServer/Persistence.md:33]:
```suggestion
Where a custom query is reused in multiple locations, we use an extension method class on the `IDocumentStore<TDocument>` interface. For example, we have a [reusable mechanism for loading a `Project` via its slug](https://github.com/OctopusDeploy/OctopusDeploy/blob/fb83083950a4bcac1d76ebcb6fc0b9250ba262b3/source/Octopus.Core/Features/Projects/ProjectDocumentStoreExtensionMethods.cs#L16).
```

File: OctopusServer/Persistence.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+In these situations, you'll need to make it clear which space and project path you mean when you ask, for instance, for an `IDocumentStore<Runbook>` to fetch a runbook. To do that, take a dependency on an `IProjectScope` and structure your code like this:
+
+```
+using (projectScope.Push(spaceId, projectPath))
+{
+    var runbook = await runbookDocumentStore.GetAsync(runbookId, cancellationToken);
+    runbook.DoTheThing();
+}
+```
+You can compose a `ProjectPath` from either just a `ProjectId` or a `ProjectId` and `GitBranch`.
```
Comment (Waldo000000) [inline OctopusServer/Persistence.md:57]: Raises the question of: When can I/should I use "just a `ProjectId`", and which branch is assumed in that case?

File: OctopusServer/Persistence.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Persistence
```
Comment (Waldo000000) [inline OctopusServer/Persistence.md:1]: 💡 Is there a natural place for this doco as a README.md right alongside the code, in https://github.com/OctopusDeploy/OctopusDeploy/tree/master/source/Octopus.Core/Persistence?

**Review by Waldo000000 — APPROVED** [review]
Nice, some really useful stuff there, easy to read too 👏 

Left some comments to consider; HTH! :shipit:

File: OctopusServer/Persistence.md (hunk 1 above)
Comment (uglybugger) [reply OctopusServer/Persistence.md:24]: Good point. I'll add a thing.

File: OctopusServer/Persistence.md (hunk 5 above)
Comment (uglybugger) [reply OctopusServer/Persistence.md:1]: If we could have a symlink, I'd have it in both :)

I think there's at least a place for a README.md in the code linking to the `Persistence.md` page in this repository.

File: OctopusServer/Persistence.md (hunk 4 above)
Comment (uglybugger) [reply OctopusServer/Persistence.md]: Good point. Addressed in 2d34ce0e5298a4027eeea62ca2b97e3a80badf69.

File: OctopusServer/Persistence.md (hunk 1 above)
Comment (uglybugger) [reply OctopusServer/Persistence.md:24]: Thing added in dd7df1464f253901473c18569b81f2fe17a5fabf.

File: OctopusServer/Persistence.md (hunk 5 above)
Comment (uglybugger) [reply OctopusServer/Persistence.md:1]: (On my list before adding that is consolidating all the persistence-related files into a `Persistence` namespace - they're a bit spread out at the moment.

File: OctopusServer/Persistence.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+# FAQ
+
+## I need to create/read/modify/delete a Foo entity. How should I do that?
+
+Take a constructor-level dependency on an `IDocumentStore<Foo>`.
+
+At present, if you modify something, you'll still need to call `await fooDocumentStore.UpdateAsync(foo, cancellationToken);` to tell the store that you've changed something. This requirement will go away (along with the `.Update(...)` and `.UpdateAsync(...)` methods). The safe assumption for now is that if those methods exist you should call them, and when they cease to exist everything will just continue to work.
+
+**IMPORTANT:** Please don't modify an entity if you don't want your changes persisted. At some point, persistence of any mutations will happen automatically, so even if you're not currently calling `.UpdateAsync(...)` the behaviour will soon change so that your changes will be picked up anyway.
```
Comment (zentron) [inline OctopusServer/Persistence.md:69]: I'd be interested to see a link out to this architectural discussion as moving away from explicit modifications has its own design discussions.

File: OctopusServer/Persistence.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+You _are_ likely to encounter explicit knowledge of project scoping in the following scenarios:
+1. Iterating over collections of projects;
+1. During background task execution; and
+1. In message bus event handlers (which implement `IEventuallyHandle<TEvent>`).
+
+In these situations, you'll need to make it clear which space and project path you mean when you ask, for instance, for an `IDocumentStore<Runbook>` to fetch a runbook. To do that, take a dependency on an `IProjectScope` and structure your code like this:
+
+```
+using (projectScope.Push(spaceId, projectPath))
```
Comment (zentron) [inline OctopusServer/Persistence.md:51]: This section probably needs a bit of a DANGER callout that it should really be an exceptional case and the implementer should understand what the dependency chain is around the using block due to the nature of modifying our (lifetime-scoped)  project context "singleton".
I'm actually thinking we should enforce the adding of comments each time this tool is used.

**Review by zentron — APPROVED** [review]
looks good. is it worth adding a section on some of the decorators around the persistence classes to explain the what, why and how?

**Review by MarkSiedle — APPROVED** [review]
Well explained 👍

File: OctopusServer/Persistence.md (hunk 7 above)
Comment (uglybugger) [reply OctopusServer/Persistence.md:53]: Added an advisory note in e0606c4a24d9b31aa386d76073dbe1bb45bc07c3.

**Pull request merged**